### PR TITLE
Defer variant resolution for jar-packaged components

### PR DESCRIPTION
**Pull request by DPUkyle**
Defer variant resolution for jar-packaged components
 
 - Only returns FixedComponentArtifacts for relocated coordinates, all other cases will derive artifacts from all available variants.
 - At face value, this PR does not significantly change any user-facing behavior. However it will enable a downstream change (along with #19784) to allow resolving ArtifactView from _any_ configuration, not the single configuration used to build the dependency graph.
 - Due to the size of the ArtifactView changes, @big-guy and I are splitting into multiple PRs.

### CODE REVIEW

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
I am lacking the context, but removing that code does not sound right and the number of failing tests seems to be a confirmation.

**Comment by big-guy** [comment]
@ljacomet we didn't expect this to work, we were just trying to separate these failures from the other spike branch.

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/VariantFilesMetadataRulesIntegrationTest.groovy [hunk 1]
```diff
@@ -665,7 +665,7 @@ class VariantFilesMetadataRulesIntegrationTest extends AbstractModuleDependencyR
         applyFileRule | extension | not
         true          | 'jar'     | ''
         false         | 'jar'     | 'not'
-        true          | 'notJar'  | ''
+        false         | 'notJar'  | ''
         false         | 'notJar'  | 'not'
```
Comment (DPUkyle) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/VariantFilesMetadataRulesIntegrationTest.groovy:669]: @big-guy the last row of this test's data table is pretty suspicious. It passes, but I'd like your opinion on whether this is still relevant.

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle help

**Comment by DPUkyle** [comment]
@bot-gradle test QF PT

**Comment by DPUkyle** [comment]
@bot-gradle test QF PT

**Comment by DPUkyle** [comment]
@bot-gradle test QF PT

**Comment by DPUkyle** [comment]
@bot-gradle test PT

**Comment by DPUkyle** [comment]
@bot-gradle test QF PT

**Comment by DPUkyle** [comment]
@bot-gradle test QF PT

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle QF

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle cancel

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle PRF PT

**Comment by DPUkyle** [comment]
@bot-gradle cancel

**Comment by DPUkyle** [comment]
@bot-gradle test PRF PT

**Comment by DPUkyle** [comment]
@bot-gradle test PRF PT

**Comment by DPUkyle** [comment]
@bot-gradle test PT

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultArtifactSet.java [hunk 2]
```diff
@@ -282,6 +282,11 @@ public void visitDependencies(TaskDependencyResolveContext context) {
         public File calculateValue(NodeExecutionContext context) {
             DefaultBuildableArtifactResolveResult result = new DefaultBuildableArtifactResolveResult();
             artifactResolver.resolveArtifact(artifact, moduleSources, result);
+            if (!result.isSuccessful() && artifact.getAlternativeArtifact().isPresent()) {
+                DefaultBuildableArtifactResolveResult alternative = new DefaultBuildableArtifactResolveResult();
+                artifactResolver.resolveArtifact(artifact.getAlternativeArtifact().get(), moduleSources, alternative);
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultArtifactSet.java:287]: I could do this, but am not sure how to test it:
```
if (!result.isSuccessful() && artifact.getAlternativeArtifact().isPresent()) {
    DefaultBuildableArtifactResolveResult alternative = new DefaultBuildableArtifactResolveResult();
+   result.applyTo(alternative);
    artifactResolver.resolveArtifact(artifact.getAlternativeArtifact().get(), moduleSources, alternative);
}
```

[the remark this answers is not in the document]
Comment (big-guy) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultArtifactSet.java:287]: I think we can ignore this for now and revisit it in a later PR

**Review by big-guy — APPROVED** [review]

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**